cms/core/blocks/embeddable.py

```diff
@@ -98,3 +99,55 @@ def clean(self, value: "StructValue") -> "StructValue":
     class Meta:
         icon = "code"
         template = "templates/components/streamfield/ons_embed_block.html"
+
+
+class VideoEmbedBlock(blocks.StructBlock):
+    """A video embed block."""
+
+    link_url = blocks.URLBlock(
+        help_text=_(
+            "The URL to the video hosted on YouTube or Vimeo, for example, "
+            "https://www.youtube.com/watch?v={ video ID } or https://vimeo.com/video/{ video ID }. "
+            "Used to link to the video when cookies are not enabled."
+        )
+    )
+    image = ImageChooserBlock(help_text=_("The video cover image, used when cookies are not enabled."))
+    title = blocks.CharBlock(help_text=_("The descriptive title for the video used by screen readers."))
+    link_text = blocks.CharBlock(
+        help_text=_("The text to be shown when cookies are not enabled e.g. 'Watch the {title} on Youtube'.")
+    )
+
+    def get_context(self, value: "StreamValue", parent_context: dict | None = None) -> dict:
+        context: dict = super().get_context(value, parent_context=parent_context)
+        embed_url = ""
+        if urlparse(value["link_url"]).hostname in ["www.vimeo.com", "vimeo.com"]:
+            video_id = urlparse(value["link_url"]).path.split("/")[-1]
+            embed_url = "https://player.vimeo.com/video/" + video_id
+        elif urlparse(value["link_url"]).hostname in ["www.youtube.com", "youtube.com"]:
+            query = dict(param.split("=") for param in urlparse(value["link_url"]).query.split("&"))
+            video_id = query.get("v", "")
+            embed_url = "https://www.youtube.com/embed/" + video_id
+        context["value"]["embed_url"] = embed_url
+        return context
+
+    def clean(self, value: "StructValue") -> "StructValue":
+        """Checks that the given embed and link urls match youtube or vimeo."""
+        errors = {}
+
+        if urlparse(value["link_url"]).hostname not in [
```

Comment:
We need a bit more validation. For example `https://www.youtube.com/embed/ID` has the valid hostname, but not the structure we are looking for (`https://www.youtube.com/watch?v=ID` or `https://youtu.be/ID`. the latter is what you get when you click share)

I'd use https://github.com/wagtail/wagtail/blob/4fc2c2abacbe0cf8f78b122fac444984590d6f35/wagtail/embeds/oembed_providers.py#L609-L611

and for Vimeo i'd verify `https://vimeo.com/ID`, `https://player.vimeo.com/ID` (slightly modified https://github.com/wagtail/wagtail/blob/4fc2c2abacbe0cf8f78b122fac444984590d6f35/wagtail/embeds/oembed_providers.py#L559-L560) and `https://vimeo.com/*/video/ID` (example https://vimeo.com/showcase/7934865/video/493407585)

Reply:
sorry, should be `https://player.vimeo.com/video/ID` (https://player.vimeo.com/video/908205163)

Reply:
Had a go at this now, and updated the code to extract the embed url according to the different valid formats. I've also added a bunch of tests for various valid and invalid urls.

Reply:
It looks like the linter doesn't like the number of methods I've added to the 'CoreBlocksTestCase` class. Investigating this now.

Reply:
Resolved by moving the VideoEmbed tests to their own class.